oximeter/db/src/oxql/plan/plan.rs

```diff
@@ -0,0 +1,1366 @@
+                    }
+                }
+            }
+        }
+
+        let out = processed_nodes.make_contiguous().to_vec();
+        if modified {
+            Ok(OptimizedPlan::Optimized(out))
+        } else {
+            Ok(OptimizedPlan::Unchanged(out))
+        }
+    }
+
+    // Push down predicates in the list of plan nodes.
+    fn pushdown_predicates(nodes: &[Node]) -> anyhow::Result<OptimizedPlan> {
+        Self::pushdown_predicates_impl(nodes, None)
+    }
+
+    // Recursive implementation of predicate pushdown.
+    fn pushdown_predicates_impl(
+        nodes: &[Node],
+        outer_predicates: Option<Predicates>,
+    ) -> anyhow::Result<OptimizedPlan> {
+        anyhow::ensure!(
+            !nodes.is_empty(),
+            "Planning error: plan nodes cannot be empty"
+        );
+
+        // Used to return correct variant of `OptimizedPlan`.
+        let mut modified = false;
+
+        // Collect the nodes in the plan.
+        //
+        // We'll process the query plan from back to front, pushing nodes onto
+        // the output plan as we consider them. This is so we can "push" the
+        // filters from the back of the plan towards the front, as we process
+        // nodes.
+        let mut remaining_nodes =
+            nodes.iter().cloned().collect::<VecDeque<_>>();
+        let mut processed_nodes = VecDeque::with_capacity(nodes.len());
+
+        // If we were provided with outer predicates, let's just add on a filter
+        // plan node to the provided plan. It will be processed like any other
+        // node.
+        if let Some(predicates) = outer_predicates {
+            let filt = Filter::from_predicates(
+                predicates,
+                nodes
+                    .last()
+                    .expect("length verified above")
+                    .output()
+                    .into_input(),
+            )?;
+            remaining_nodes.push_back(Node::Filter(filt));
+        };
+
+        // Process nodes in the query plan, in reverse.
+        while let Some(current_node) = remaining_nodes.pop_back() {
+            // What we do with the node depends on what's in front of it, so
+            // take the next node in the plan.
+            let Some(next_node) = remaining_nodes.pop_back() else {
+                // If there _isn't_ one, then the current node must be the start
+                // of the plan, and so push it and break out.
+                processed_nodes.push_front(current_node);
+                break;
+            };
+
+            // If the current node isn't a filter, then this pass doesn't modify
+            // it at all. Instead, just push it and continue. But! We have to
+            // push back the next node we just popped above.
+            let Node::Filter(current_filter) = current_node else {
+                processed_nodes.push_front(current_node);
+                remaining_nodes.push_back(next_node);
```

Comment:
See the related comment in the corresponding section of `pushdown_limit_impl()`, but I think we may be able to push this onto `processed_nodes`, iff it's not a filter operation. Or we can just leave this, and always push it onto the remaining nodes, I think that's correct too.